### PR TITLE
Complete refactor of the Emulator

### PR DESCRIPTION
**Pull request by jmdobry**
__tldr;__ Refactored the Emulator server into a REST/gRPC Service implementing the Cloud Functions API as defined in [https://cloudfunctions.googleapis.com/$discovery/rest?version=v1beta2](https://cloudfunctions.googleapis.com/$discovery/rest?version=v1beta2) and [functions.proto](https://github.com/googleapis/googleapis/blob/master/google/cloud/functions/v1beta2/functions.proto). 

The Emulator has two services: REST and gRPC. The CLI can be made to talk to either one of them. By default it talks to the REST service. The Google Cloud SDK can be configured to interact with the REST service of the Emulator instead of the production Cloud Functions service with: `gcloud config set api_endpoint_overrides/cloudfunctions http://localhost:8008` 

A new Supervisor service manages the invocation of locally deployed functions. It has two isolation models: `inprocess` and `childprocess`. `inprocess` obviously carries the risk that the Supervisor process itself could get borked or frozen by a user's function. The debugging workflow with `childprocess` is more cumbersome, but the invocation of the user's function is isolated in its own process.

A set of Models allow for easy data management for functions and operations. It uses the Adapter Pattern, but currently only has one adapter, `ConfigAdapter`, which uses the `configstore` library to store data in the user's home directory.

* [x] - README is completely up-to-date
* [x] - GrpcClient
* [x] -  RestClient
* [x] -  GrpcService
* [x] -  RestService
* [x] -  Models
* [x] -  Proto utils
* [x] -  New Supervisor
* [x] - Tests

### CODE REVIEW

**Comment by jmdobry** [comment]
@jasonpolites Want to do a final review? This is pretty much done. I'd like to merge it and then start using issues/milestones/projects to track bugs/features. This PR is getting unwieldy.

File: README.md [hunk 1]
```diff
@@ -102,37 +106,42 @@ Print the available commands:
...
+                               module.
+      restart                  Restarts the emulator.
+      start                    Starts the emulator.
+      status                   Reports the current status of the emulator.
+      stop                     Stops the emulator gracefully.
 
     Options:
-      --host, -h  The emulator's host.                                                                              [string]
-      --port, -p  The emulator's port.                                                                              [number]
+      --grpcHost  The host the gRPC Service should run on.                                                          [string]
```
Comment (jasonpolites) [inline README.md]: [nit]  "The host on which the gRPC Service should run. "

File: src/cli/commands/status.js [hunk 2]
```diff
@@ -35,25 +36,81 @@ exports.handler = (opts) => {
...
+          } else if (time > 60) {
+            time = `${Math.floor(time / 60)} minute(s)`;
+          } else {
+            time = `${Math.floor(time)} second(s)`;
+          }
+        }
+
+        table.push(['Status'.white, 'STOPPED'.yellow]);
+        if (time) {
+          table.push(['Last up'.white, `${time.yellow} ago`]);
```
Comment (jasonpolites) [inline src/cli/commands/status.js]: This results in 'ago' being "default console text color", and the time being in yellow.. which looks a little odd to me

File: README.md [hunk 3]
```diff
@@ -322,6 +349,12 @@ debugger window. You can then set breakpoints and debug like normal.
 
 ![Debugging with Chrome Developer Tools](img/debugging.png "Debugging with Chrome Developer Tools")
 
+#### Debugging with WebStorm
+
+Create a Node.js Remote Debugger configuration. In the Emulator, set `debug` to
+`true`. When your function is invoked and then paused, attach to the debugger in
```
Comment (jmdobry) [inline README.md]: Added clarification.

File: README.md (hunk 1 above)
Comment (jmdobry) [inline README.md]: Fixed
